packages/core/nano/src/types.ts

```diff
@@ -1,22 +1,139 @@
-export interface Consumer<T> {
-    notify: <K extends keyof T>(property: K, value: T[K]) => void;
-    notifyMany: <K extends keyof T>(values: Pick<T, K>) => void;
+export interface Pending<T> {
+  kind: "Pending";
+  estimate?: T;
 }
 
-export interface Pending<T> {
-    kind: "Pending";
-    estimate?: T;
+/**
+ * The interface for an object whose data can be bound to. We use this contract for
+ * components that want to expose their data and its changes to other components.
+ * An example of this would be Formula binding to a Tablero. In this scenario,
+ * Tablero is expected to implement IProducer and expose some of its data, given
+ * a reference, through getDataResult. Formula is expected to implement IConsumer
+ * and register itself as an observer to Tablero.
+ *
+ * Any component that implements IProducer is expected to provide some registration
+ * functionality and to notify observers whenever the data they are bound to changes.
+ */
+export interface IProducer<T> {
+  /**
+   * Binds the given observer to the component associated with this Producer.
+   * @param consumer - The object to start listening to any changes in the bound data.
+   */
+  addConsumer(consumer: IConsumer<T>): void;
+
+  /**
+   * Unbinds the given observer from the component associated with this Producer.
+   * @param consumer - The observer to unregister from the Producer.
+   */
+  removeConsumer(consumer: IConsumer<T>): void;
+
+  /**
+   * Return IDataResult for the given dataBindingReference, containing the data value
+   * and some additional metadata.
+   * @param dataBindingReference - An instance of dataBindingReference, a reference
+   * to some data in the bound component.
+   */
+  read<K extends keyof T>(consumer: IConsumer<T>, property: K, ...args: any[]): T[K] | Pending<T[K]>;
+
+  /**
+   * Invoked when this data provider is deleted.
+   */
+  onDeleted(): void;
```

Comment:
What is the purpose of the 'onDeleted()' callback?

In other systems I've work on, a producer releases resources when the number of consumers drops to zero.  Would that work for your scenario?

Reply:
How are consumers notified when a producer is deleted? In a more practical example, say we have a Formula bound to the name of a table. When the table is deleted, we don't want the formula to show the table name anymore, we probably want it to show some broken reference UI. Also, we'd want to make sure the producer is unregistered from the DataBindingService (not a service we're adding into this repo). Can those concepts still be addressed without onDeleted?